src/lib/mina_base/parties_logic.ml

```diff
@@ -0,0 +1,303 @@
+           ; inclusion_proof })
+    in
+    let party_succeeded = Bool.( && ) predicate_satisfied update_permitted in
+    let local_state =
+      { local_state with
+        success= Bool.( && ) local_state.success party_succeeded }
+    in
+    let local_delta =
+      Amount.(h.perform (Balance a) - h.perform (Balance a'))
+    in
+    let party_token = h.perform (Party_token_id party) in
+    let fee_excess_change0, new_local_fee_excess =
+      let curr_token : Token_id.t = local_state.token_id in
+      let same_token = Token_id.equal curr_token party_token in
+      let curr_is_default = Token_id.(equal default) curr_token in
+      let should_merge = Bool.( && ) (Bool.not same_token) curr_is_default in
+      let to_merge_amount =
+        Amount.if_ should_merge ~then_:local_state.excess ~else_:Amount.zero
+      in
+      let new_local =
+        let base =
+          Amount.if_ same_token ~then_:local_state.excess ~else_:Amount.zero
+        in
+        Amount.add_signed base local_delta
+      in
+      (to_merge_amount, new_local)
+    in
+    let local_state = {local_state with excess= new_local_fee_excess} in
+    let global_state =
+      (* TODO: Maybe overflows should be possible and cause a transaction failure? *)
+      h.perform
+        (Modify_global_excess (global_state, Amount.( + ) fee_excess_change0))
+    in
+    (* If a's token ID differs from that in the local state, then
+      the local state excess gets moved into the execution state's fee excess.
+
+      If there are more parties to execute after this one, then the local delta gets
+      accumulated in the local state (unless there is a token
+
+      If there are no more parties to execute, then we do the same as if we switch tokens.
+      The local state excess (plus the local delta) gets moved to the fee excess if it is default token.
+    *)
+    let new_ledger =
+      let should_apply = Bool.( || ) is_start local_state.will_succeed in
+      h.perform
+        (Set_account_if (should_apply, local_state.ledger, a', inclusion_proof))
+    in
+    let local_state = {local_state with ledger= new_ledger} in
+    let is_last_party = Ps.is_empty remaining in
+    h.perform (Finalize_local_state (is_last_party, local_state)) ;
+    (*
+       In the SNARK, this will be
+    Bool.(assert_
+            (not is_last_party ||
+            equal local_state.will_succeed local_state.success)) ;
+       *)
+    let global_state =
+      let curr_is_default = Token_id.(equal default) local_state.token_id in
+      let should_perform_second_excess_merge =
+        (* The first party's fee always gets merged in *)
+        Bool.(curr_is_default && (is_last_party || is_start))
+      in
+      h.perform
+        (Modify_global_excess
+           ( global_state
+           , fun amt ->
+               Amount.if_ should_perform_second_excess_merge
+                 ~then_:Amount.(amt + local_state.excess)
+                 ~else_:amt ))
+    in
```

Comment:
I think you'll want to set the `local_state.excess` here to `Amount.zero` if `should_perform_second_excess_merge`.